### PR TITLE
MVP Actions in v2

### PR DESCRIPTION
**Pull request by ericanderson**

### CODE REVIEW

File: examples/basic/sdk/src/generatedNoCheck/Ontology.ts [hunk 1]
```diff
@@ -1,18 +1,25 @@
...
-    Employee: Employee,
-    ObjectTypeWithAllPropertyTypes: ObjectTypeWithAllPropertyTypes,
+    Todo: Objects.Todo,
+    Person: Objects.Person,
+    Employee: Objects.Employee,
+    ObjectTypeWithAllPropertyTypes: Objects.ObjectTypeWithAllPropertyTypes,
+  },
+  actions: {
+    actionTakesAllParameterTypes: Actions.actionTakesAllParameterTypes,
+    createTodo: Actions.createTodo,
```
Comment (mfedderly) [inline examples/basic/sdk/src/generatedNoCheck/Ontology.ts:16]: This is a nice way of avoiding collisions.

File: packages/client/src/actions/executeAction.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  actionApiName: A,
+  parameters?: Parameters<Actions<O>[A]>[0],
+  options?: Op,
+): Promise<void> {
+  const response = await applyActionV2(
+    createOpenApiRequest(client.stack, client.fetch),
+    client.ontology.metadata.ontologyApiName,
+    actionApiName as string,
+    {
+      parameters: parameters as any,
```
Comment (mfedderly) [inline packages/client/src/actions/executeAction.ts:39]: Can we actually get away without having to marshal all the types to the wire format like in v1.1?

File: packages/client/src/types/Actions.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+type ActionParameterKeysFrom<
+  O extends OntologyDefinition<any>,
+  K extends ActionKeysFrom<O>,
+> = keyof ActionDefinitionFrom<O, K>["parameters"];
+
+type ActionParameterDefinitionFrom<
+  O extends OntologyDefinition<any>,
+  K extends ActionKeysFrom<O>,
+  P extends keyof O["actions"][K]["parameters"],
```
Comment (mfedderly) [inline packages/client/src/types/Actions.ts:46]:
```suggestion
  P extends ActionParameterKeysFrom<O, K>,
```

File: packages/client/src/types/Actions.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+> = ActionDefinitionFrom<O, K>["parameters"][P];
+
+type ActionParameterTypeFrom<
+  O extends OntologyDefinition<any>,
+  K extends ActionKeysFrom<O>,
+  P extends keyof O["actions"][K]["parameters"],
```
Comment (mfedderly) [inline packages/client/src/types/Actions.ts:52]: `ActionParameterKeysFrom` again

File: packages/client/src/types/Actions.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+> = NullableProps<ActionDefinitionFrom<O, K>["parameters"]> extends never
+  ? OsdkActionParametersNotOptional<O, K>
+  : PartialByNotStrict<
+    OsdkActionParametersNotOptional<O, K>,
+    NullableProps<ActionDefinitionFrom<O, K>["parameters"]>
+  >;
+
+export type Actions<O extends OntologyDefinition<any>> = {
+  [K in ActionKeysFrom<O>]: (
+    args: NOOP<OsdkActionParameters<O, K>>,
```
[the remark this answers is not in the document]
Comment (mfedderly) [reply packages/client/src/types/Actions.ts:101]: We should probably also hand the no arguments case here as well. I'm assuming there's actions that don't actually have any parameters?

File: packages/client/src/actions/executeAction.ts (hunk 2 above)
Comment (pashakondratyev) [reply packages/client/src/actions/executeAction.ts:39]: Yeah we will need to handle these, but I don't think we've decided on the what attachments, timestamps, and dates will look like yet entirely. Objects and arrays of Objects wont work either I believe, and we need a way to extract the Object Set definition here as well. All the other primitives should be fine

Comment (ericanderson) [reply packages/client/src/actions/executeAction.ts:39]: For most things I was just going to lean into raw types as long as possible. Good points on the edge cases

**Pull request merged**